pkg/testing/multipass/provisioner.go

```diff
@@ -0,0 +1,229 @@
+	// multipass cannot handle it, it either results in instances sharing the same IP address
+	// or some instances stuck in Starting state
+	for _, batch := range batches {
+		err := func(batch runner.OSBatch) error {
+			launchCtx, launchCancel := context.WithTimeout(ctx, 5*time.Minute)
+			defer launchCancel()
+			err := p.launch(launchCtx, batch)
+			if err != nil {
+				return fmt.Errorf("instance %s failed: %w", batch.ID, err)
+			}
+			return nil
+		}(batch)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var results []runner.Instance
+	instances, err := p.list(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, batch := range batches {
+		mi, ok := instances[batch.ID]
+		if !ok {
+			return nil, fmt.Errorf("failed to find %s in multipass list output", batch.ID)
+		}
+		if mi.State != "Running" {
+			return nil, fmt.Errorf("instance %s is not marked as running", batch.ID)
+		}
+		results = append(results, runner.Instance{
+			ID:         batch.ID,
+			Name:       batch.ID,
+			IP:         mi.IPv4[0],
+			Username:   "ubuntu",
+			RemotePath: "/home/ubuntu/agent",
+			Internal:   nil,
+		})
+	}
+	return results, nil
+}
+
+// Clean cleans up all provisioned resources.
+func (p *provisioner) Clean(ctx context.Context, instances []runner.Instance) error {
+	// doesn't execute in parallel for the same reasons in Provision
+	// multipass just cannot handle it
+	for _, instance := range instances {
+		func(instance runner.Instance) {
+			deleteCtx, deleteCancel := context.WithTimeout(ctx, 5*time.Minute)
+			defer deleteCancel()
+			err := p.delete(deleteCtx, instance)
+			if err != nil {
+				// prevent a failure from stopping the other instances and clean
+				p.logger.Logf("Delete instance %s failed: %s", instance.Name, err)
+			}
+		}(instance)
+	}
+	return nil
+}
+
+// launch creates an instance.
+func (p *provisioner) launch(ctx context.Context, batch runner.OSBatch) error {
+	args := []string{
+		"launch",
+		"-c", "2",
+		"-d", "50G", // need decent size for all the tests
+		"-m", "4G",
+		"-n", batch.ID,
+		"--cloud-init", "-",
+		batch.OS.Version,
+	}
+
+	publicKeyPath := filepath.Join(".integration-cache", "id_rsa.pub")
+	publicKey, err := os.ReadFile(publicKeyPath)
+	if err != nil {
+		return fmt.Errorf("failed to read SSH key to send to multipass instance at %s: %w", publicKeyPath, err)
+	}
+
+	var cfg cloudinitConfig
+	cfg.SSHAuthorizedKeys = []string{string(publicKey)}
+	cfgData, err := yaml.Marshal(&cfg)
+	if err != nil {
+		return fmt.Errorf("failed to marshal cloud-init configuration: %w", err)
+	}
+
+	var output bytes.Buffer
+	p.logger.Logf("Launching multipass image %s", batch.ID)
+	proc, err := process.Start("multipass", process.WithContext(ctx), process.WithArgs(args), process.WithCmdOptions(runner.AttachOut(&output), runner.AttachErr(&output)))
+	if err != nil {
+		return fmt.Errorf("failed to run multipass launch: %w", err)
+	}
+	_, err = proc.Stdin.Write([]byte(fmt.Sprintf("#cloud-config\n%s", cfgData)))
```

Comment:
So, is the multipass config using `cloud-config` for local VMs? Is so, the ignore the comment above.

Reply:
Yes, you will find this being used by all distros to provide a way of configuring a new instance.